src/live_restore/live_restore_fs.c

```diff
@@ -597,72 +480,54 @@ typedef enum { NONE, FULL, PARTIAL } WT_LIVE_RESTORE_SERVICE_STATE;
-    while (hole != NULL) {
-        if (read_end < hole->off)
-            /* All subsequent holes are past the read. We won't find matching holes. */
-            break;
-
-        WT_ASSERT_ALWAYS(session, !(offset < hole->off && WTI_EXTENT_END(hole) < read_end),
-          "Read (offset: %" PRId64 ", len: %" WT_SIZET_FMT ") encompasses a hole (offset: %" PRId64
-          ", len: %" WT_SIZET_FMT ")",
-          offset, len, hole->off, hole->len);
-
-        read_begins_in_hole = WTI_OFFSET_IN_EXTENT(offset, hole);
-        read_ends_in_hole = WTI_OFFSET_IN_EXTENT(read_end, hole);
-        if (read_begins_in_hole && read_ends_in_hole) {
-            /* Our read is entirely within a hole */
-            __wt_verbose_debug3(session, WT_VERB_LIVE_RESTORE,
-              "CANNOT SERVICE %s: Reading from hole. Read: %" PRId64 "-%" PRId64 ", hole: %" PRId64
-              "-%" PRId64,
-              lr_fh->iface.name, offset, read_end, hole->off, WTI_EXTENT_END(hole));
-            return (NONE);
-        } else if (!read_begins_in_hole && read_ends_in_hole) {
-            /*
-             * The block manager reads entire pages so we can expect all reads to exist entirely
-             * inside or outside a hole during normal WiredTiger operation. The one exception is
-             * when background migration threads are running as they will copy data chunks
-             * regardless of page size. The background threads always migrate a file from start to
-             * finish so the one case where we partially read from a hole is when the background
-             * thread reads the first part of a page and then we read that page before the remainder
-             * is migrated.
-             */
-            __wt_verbose_debug3(session, WT_VERB_LIVE_RESTORE,
-              "PARTIAL READ %s: Reading from hole. Read: %" PRId64 "-%" PRId64 ", hole: %" PRId64
-              "-%" PRId64,
-              lr_fh->iface.name, offset, read_end, hole->off, WTI_EXTENT_END(hole));
-            *holep = hole;
-            return (PARTIAL);
-        } else if (read_begins_in_hole && !read_ends_in_hole) {
-            /* A partial read should never begin in a hole. */
-            WT_ASSERT_ALWAYS(session, false,
-              "Read (offset: %" PRId64 ", len: %" WT_SIZET_FMT
-              ") begins in a hole but does not end in one (offset: %" PRId64
-              ", "
-              "len: %" WT_SIZET_FMT ")",
-              offset, len, hole->off, hole->len);
+    uint64_t read_end_bit = WTI_OFFSET_TO_BIT(WTI_OFFSET_END(offset, len));
+    uint64_t read_start_bit = WTI_OFFSET_TO_BIT(offset);
+    bool read_begins_in_hole = false, read_ends_in_hole = false, hole_bit_set = false;
+
+    for (uint64_t current_bit = read_start_bit; current_bit < read_end_bit; current_bit++) {
+        if (__bit_test(lr_fh->destination.bitmap, current_bit)) {
+            if (read_begins_in_hole)
+                /* A partial read should never begin in a hole. */
```

Comment:
```suggestion
                /* If we are here, we started our read in a hole but have found part of our read region already filled in. If we start our read in a hole we should expect the entire read to be within a hole. */
```
Feel free to take some or none of this but more explanation here would be great!

Reply:
This is explained in the function comment so for now I will leave it.